blocks/main_loop.py

```diff
@@ -42,6 +45,87 @@
 a model. Check documentation of your extensions."""
 
 
+class Timer(object):
+    def __init__(self):
+        self.total = defaultdict(int)
+        self.current = []
+        self.order = OrderedSet()
+
+    def enter(self, name):
+        self.current.append(name)
+        # We record the order in which sections were first called
+        self.order.add(tuple(self.current))
+
+    def exit(self, t):
+        self.total[tuple(self.current)] += t
+        self.current.pop()
+
+    def report(self):
+        """Print a report of timing information to standard output.
+
+        .. todo::
+
+           This method could accept different I/O objects (e.g. standard
+           error or write to a file).
+
+        """
+        total = sum(v for k, v in self.total.items() if len(k) == 1)
+
+        def print_report(keys, level=0):
+            subtotal = 0
+            for key in keys:
+                if len(key) > level + 1:
+                    continue
+                subtotal += self.total[key]
+                section = ' '.join(key[-1].split('_'))
+                section = section[0].upper() + section[1:]
+                print('{:30}{:15.2f}{:15.2%}'.format(
+                    level * '  ' + section, self.total[key],
+                    self.total[key] / total
+                ))
+                children = [k for k in keys
+                            if k[level] == key[level] and
+                            len(k) > level + 1]
+                child_total = print_report(children, level + 1)
+                if children:
+                    print('{:30}{:15.2f}{:15.2%}'.format(
+                        (level + 1) * '  ' + 'Other',
+                        self.total[key] - child_total,
+                        (self.total[key] - child_total) / total
+                    ))
+            return subtotal
+
+        print('{:30}{:>15}{:>15}'.format('Section', 'Time', '% of total'))
+        print('-' * 60)
+        print_report(self.order)
+
+
+class TimeIt(object):
+    """A context manager to time the execution time of code within it.
+
+    Parameters
+    ----------
+    name : str
+        The name of this section. Expected to adhere to variable naming
+        styles.
+    timer : :class:`Timer`
+        The timer of the main loop. This is the object this context manager
+        will report the execution time to. The accumulation and processing
+        of timing information is handled by this object.
+
+    """
+    def __init__(self, name, timer):
+        self.name = name
+        self.timer = timer
+
+    def __enter__(self):
+        self.timer.enter(self.name)
+        self.start = time.clock()
+
+    def __exit__(self, *args):
+        self.timer.exit(time.clock() - self.start)
```

Comment:
You are using `time.clock()`, which is CPU time. I am slightly worried if it always correctly includes time spent on waiting for a GPU operation to finish. Do you know if it's accounted?

Also I think documentation should mention this choice.